\chapter{Design \& Implementation}

\section{Initialization}
Since iHMMuneAlign is rather configurable, it requires a large amount of input data for each run that cannot be embedded in the code directly. This data has to be read during program initialisation and stored in memory such that it can be efficiently used in their required of the algorithm.

\subsection{IGHD/IGHJ Repertoires}
The use-case for the germline IGHD and IGHJ repertoires are farily straight forward. For each nucleotide in each gene, a state needs to be created that represents the probability that gene caused it to be there.

Since this needs to be done for each gene/nucleotide in order, there's no need for random access, or efficient searching. They're stored in a contiguous block of memory in a dynamic (variable size) array. This provides good cache/memory locality \autocite{cache-locality}, while being easy to use.

\subsection{IGHV Repertoire}
\label{sec:v-repertoire}
Unlike the other two genes, not every IGHV gene needs to be in the model. Instead, a BLAST search is performed, and (for now, at least) only the most likely hit --the closest germline IGHV gene-- is used.

However, the BLAST results do not in fact give the sequence in its output; It only presents the relevant matching/aligned section of the sequence. Fortunately, it also gives the (unique) name of the gene, as specified in the repertoire file it was given.

Consequently, an index must be built from the gene repertoire that maps each gene name to its corresponding sequence. Considering this lookup needs to be done for each model, it should be as efficient as possible. This index is currently stored as a hash-table in memory, allowing for efficient, constant time lookup \autocite{hash-table}.

\subsection{Mutation Ratios}
A large part of what iHMMuneAlign attempts to model is mutation of various germline genes, during the recombination process. While it is known that some nucleotides are more likely to mutate to a certain one than others, determining this is out of scope for iHMMuneAlign \autocite{iHMMuneAlign}. As a result, iHMMuneAlign takes in these probabilities as input.

A simple map from a nucleotide to nucleotide would not be sufficient, as it depends on the two surrounding nucleotides as well. These ratios are given in a format specific to the previous implementation of iHMMuneAlign; an example can be seen in \autoref{fig:mutation-probs}.

\begin{verbbox}
A(A->C)A 	0.200711181
A(A->C)C 	0.234567901
A(A->C)G 	0.090360982
A(A->C)T 	0.249740516
A(A->G)A 	0.698044251
A(A->G)C 	0.679012346
A(A->G)G 	0.798111707
A(A->G)T 	0.533465028
A(A->T)A 	0.101244567
A(A->T)C 	0.086419753
\end{verbbox}

\begin{figure}
    \centering
    \theverbbox
    \caption{An extract of a mutation probabilities file}
    \label{fig:mutation-probs}
\end{figure}

The probabilities are parsed into double-precision floating point numbers; the string on the left is split up, with all the unnecessary characters (arrow and brackets) dropped. Doing so allowed for more efficient hashing when put into a hash-table in memory, effectively forming an index much like the \nameref{sec:v-repertoire}.

\section{BLAST}
The first step is to determine the most likely germline V gene is in the target sequence. Fortunately, this can be determined quite efficiently by aligning the input sequence with all known V genes \cite{iHMMuneAlign}. For that we opted to use the excellent alignment tool by NCBI: BLAST \cite{blast}, like the original implementation.

BLAST is distributed as source code, and can be used in two ways:
\subsection{Invoking Binary}
\label{sec:blast-bin}
There are binaries for all the common tasks, that take in sequence repertoires as files in FASTA \autocite{fasta} format. The previous implementation used the \code{blastn} executable, which takes in a sequence in FASTA format, and writes the results (matching sequences in the repertoires) out to an XML file.

The primary advantage of using blast in this manner would be the ease of development. Since the binaries come pre-packaged, very little extra code would be required to invoke the binary from within iHMMuneAlign, and parse the XML output. However, doing so incurs a (potentially significant) performance penalty at run time due to the cost of:

\begin{description}
	\item[Process Creation]
	The operating system has to create a new process for the blast executable, perform all its bookkeeping operations, and clean up after it is completed.
	\item[Reading Input Files]
	The executable would then have to read the database files from disk again, instead of using the copy that already in memory (residing in iHMMuneAlign's address space).
	\item[Serialising Results]
	BLAST, like most high performance programs, uses its own internal intermediate representation for the input and germline genes\autocite{blast}. Once the alignment search is completed, it will then need to convert that data into a human readable, or standardised format that can be parsed by iHMMuneAlign.
	\item[Transferring Results]
    Modern operating systems provide a certain level of isolation, such that a process can not directly access data in a section of memory belonging to another process \autocite{address-spaces}. As a result, the previous implementation required the BLAST executable to write its results to disk so it could be read.
    
    While IO operations are generally expensive, ones involving disk access tend to be especially so \autocite{os-io}. This is due to the fact that disks are several orders of magnitude slower than main memory (RAM).
\end{description}

As explained in \secref{sec:threadbench}, these overheads were partially responsible for large performance inefficiencies in the original implementation. 

\subsection{Linking Shared Library}
\label{sec:blast-lib}
BLAST+ can also be compiled as a library, that can be statically or dynamically linked the target program using it. This would the alignment search to be performed directly in the process of iHMMuneAlign, effectively eliminating all of the overheads listed in the previous section.

The disadvantage of this approach is that it requires extra work during the development of iHMMuneAlign. This includes the work required for:
\begin{description}
	\item[Acquiring Libraries]
	The BLAST+ binary executables are quite prevalent in bioinformatics research and as such, they are readily available for most research/development platforms. The libraries however, are much less common, as they are generally used by developers working on production tools relating to BLAST+. 
	
	This would require researching the build process used by the BLAST+ libraries. This is a non-trivial task, as the documentation is somewhat lacking, and makes implicit assumptions about the build environment which do not necessarily hold outside of NCBI lab machines.
	\item[Programming for Libraries]
	\label{subsec:blast-lib-xml}
	The binaries accept the input data and germline repertoires in FASTA format, and it outputs in XML. Both of these are widely used, standard formats; There are a range of utilities in most languages to deal with them. As a result, very little code would need to be written specifically for it.
	
	Using the data-structures that comprise the internal representations used by the BLAST directly, would be much more complicated. And once again, the documentation regarding were not as comprehensive as they could have been.
	\item[Distributing Libraries]
	The binaries are quite widespread, well known, and can be assumed to be already available on machines that iHMMuneAlign would run on. Thus, it would be the users' responsibility to ensure they have it installed on their machine, as it's a required dependency.
	
	That same assumption can not be made for the library, and as a result we must distribute it with the program. Also --and more significantly--, the libraries must be available to the compiler when iHMMuneAlign is being built; This would significantly increase the work required to port iHMMuneAlign to a new platform, as well as making it infeasible to distribute primarily as source code.
\end{description}

Due to the performance considerations described in \secref{sec:blast-bin}, the we opted to link to the BLAST libraries, and perform the alignment search directly in iHMMuneAlign. However, after spending a significant amount of time attempting to overcome the difficulties detailed in \secref{sec:blast-lib}, we elected to use the binaries for now. 

At the time of writing, the current implementation of iHMMuneAlign assumes the \code{blastn} binary has already been invoked, and expects to be given its results in XML format. This process can be made into a script easily, and will be revisited at a later date, once more impactful performance optimisations have been exhausted.

\section{Pipeline}

% Notes: use autosize option to get more vertical space
% 		 use texlbl="\tex" instead of label="words"
\begin{figure}
	\label{fig:pipeline}
	\caption{Overview of the Model Solving `Pipeline'}
	\centering
	\begin{dot2tex}
		digraph G {

			{
				1[label="calculateAScore"];
				2[label="buildModel"];
				3[label="doViterbi"];
				4[label="generateResult"];
			}

			{
				node[color=none];
				"Input Sequence" -> 1;
				"BLAST Hit" -> 1;

				1 -> "A Score" -> 2;

				2 -> "States" -> 3;
				2 -> "Transitions" -> 3;

				3 -> "Most Likely Path" -> 4;
				4 -> "Output";
			}

		}
	\end{dot2tex}
\end{figure}

The entire process of creating a HMM to represent \igh recombination and solving it using Viterbi's algorithm can be split up into several smaller problems. These can be viewed as stages in an assembly line where the result of each one feeds into the next, terminating with the final result. In computing, this is called a pipeline.

\subsection{Parsing Blast Results}
\lstset{basicstyle=\ttfamily\footnotesize,breaklines=true}
\begin{figure}
	\caption{An extract from a BLAST+ output file}
	\label{fig:blast-output}
	\lstinputlisting[language=XML]{blastOutput-min.xml}
	\small Note that most lines have been removed, to highlight the important parts of the file. Sections that have been removed are marked with ellipses.
\end{figure}

At this stage, iHMMuneAlign does not integrate with the BLAST+ libraries, nor does it invoke the binaries directly; this is assumed to have been performed by the user or a start-up script. Consequently, the first stage of the pipeline becomes to parse and convert all required information from BLAST+ into an internal format, allowing for efficient model generation.

The BLAST+ binaries output their results in XML format. An extract of such an output file is shown in \autoref{fig:blast-output}. This file was generated by running \code{blastn} on the AJ512650.1 example sequence that was provided with iHMMuneAlign: ``\emph{Homo sapiens partial mRNA for immunoglobulin heavy chain variable region (IGHV gene), clone 43}''.

It contains various meta-data about the program, information about input and configurations, and details of a few best matching sequences (hits). We only care about the highest scoring hit; specifically the name of the germline gene, the start of the aligned match.

As mentioned in \secref{subsec:blast-lib-xml}, XML is a widely used format, supported by many libraries and frameworks in C++. 
\secref{sec:libs} proposed the use of C++ frameworks such as Qt\autocite{qt} and Boost\autocite{boost}; both of which support reading and writing XML files. 

They are both large and extensive frameworks that are hard to distribute, much like BLAST+. Unlike the BLAST+ libraries or the NCBI C++ toolkit however, they widely used and as a result, much more readily available on lab machines that iHMMuneAlign will be used in. Nevertheless, their adoption was delayed until absolutely necessary; in an effort to make iHMMuneAlign easy to distribute.

Presently, a small and lightweight, header-only library called PugiXML\cite{pugixml} is used to parse the \code{blastn} output file. It very efficient, and sufficiently easy to use; but its continued use will be re-evaluated should iHMMuneAlign becomes dependent on Boost or Qt.

\subsection{Calculating A-Score}
\subsection{Generating Model}
\subsubsection{Generating V/D/J States}
\subsubsection{Generating Transitions}
\subsection{Running Viterbi}
\subsection{Printing Results}


